File: dp/0012_max_product/maxprod.go
package maxproduct

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maxProduct(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	maxProd, minProd, result := nums[0], nums[0], nums[0]
	for i := 1; i < len(nums); i++ {
		if nums[i] < 0 {
			maxProd, minProd = minProd, maxProd
		}
		maxProd = max(nums[i], nums[i]*maxProd)
		minProd = min(nums[i], nums[i]*minProd)
		result = max(result, maxProd)
	}
	return result
}

File: dp/0012_max_product/maxprod.cc
#include <vector>
#include <algorithm> // For std::max and std::min

int maxProduct(std::vector<int>& nums) {
    if (nums.empty()) return 0;

    int maxProd = nums[0], minProd = nums[0], result = nums[0];
    for (int i = 1; i < nums.size(); ++i) {
        if (nums[i] < 0) std::swap(maxProd, minProd);
        maxProd = std::max(nums[i], nums[i] * maxProd);
        minProd = std::min(nums[i], nums[i] * minProd);
        result = std::max(result, maxProd);
    }
    return result;
}
